### PR TITLE
Accept Seq, MutableSeq objects in Seq, MutableSeq initializers ...

### PR DESCRIPTION
**Pull request by mdehoon**
In the `__init__` method of `Seq` and `MutableSeq`, allow the argument to be `Seq` or `MutableSeq` so we can get rid of the `toseq`, `tomutable` methods.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Closes #...

### CODE REVIEW

File: Bio/Seq.py [hunk 1]
```diff
@@ -1700,16 +1705,23 @@ class MutableSeq:
...
+            if data.typecode != "u":
+                raise ValueError(
+                    "data should be a string, array of characters, Seq object, "
+                    "or MutableSeq object"
+                )
+            self._data = data
+        elif isinstance(data, str):  # TODO - What about unicode?
             self._data = array.array("u", data)
-        elif isinstance(data, (Seq, int, float)):
+        elif isinstance(data, (Seq, MutableSeq)):
```
Comment (peterjc) [inline Bio/Seq.py:1718]: Can't you use a copy of the input MutableSeq's array directly? It seems wasteful to call str on it.

**Comment by peterjc** [comment]
One query about the implementation, but what is your motivation? Why do you want to get rid of ``.toseq()`` and ``.tomutableseq()``?

My concern is making the ``__init__`` more flexible blurs the distinction between strings and biological sequences, making type inference harder, especially when reading other people's code.

Related to that, I think it will also break the type annotation work on #3165 - i.e. make more work to get that merged.

**Comment by mdehoon** [comment]
> One query about the implementation, but what is your motivation? Why do you want to get rid of `.toseq()` and `.tomutableseq()`?

Because they are not needed.

> My concern is making the `__init__` more flexible blurs the distinction between strings and biological sequences, making type inference harder, especially when reading other people's code.

Sorry, I don't understand. Usually it is possible to initialize an object with an object of the same class:
```
>>> int(7)
7
>>> tuple((3,3,4,))
(3, 3, 4)
>>> list([3,3,4,])
[3, 3, 4]
>>> str("ABCD")
'ABCD'
```
So it's quite surprising that `Seq` and `MutableSeq` specifically disallow that.

> Related to that, I think it will also break the type annotation work on #3165 - i.e. make more work to get that merged.

Can you give a specific example where things break?

**Comment by peterjc** [comment]
For the type annotation, rather than:

```python
def __init__(self, data: str) -> None
```

we'd need to say ``data`` can be str, Seq or MutableSeq. Not the end of the world, but the longer #3165 waits there more of this kind of tweaking will be needed.

**Comment by peterjc** [comment]
Looks like the last change needs to consider the DBSeq too,

```
======================================================================
ERROR: test_convert (common_BioSQL.SeqInterfaceTest)
Check can turn a DBSeq object into a Seq or MutableSeq.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\biopython\Tests\common_BioSQL.py", line 504, in test_convert
    other = MutableSeq(test_seq)
  File "C:\projects\biopython\build\lib.win32-3.8\Bio\Seq.py", line 1719, in __init__
    self._data = array.array("u", data._data)
AttributeError: 'DBSeq' object has no attribute '_data'
======================================================================
ERROR: test_Seq_objs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_tests.py", line 275, in runTest
    suite = loader.loadTestsFromName(name)
  File "C:\Py\lib\unittest\loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "C:\projects\biopython\Tests\test_Seq_objs.py", line 70, in <module>
    class StringMethodTests(unittest.TestCase):
  File "C:\projects\biopython\Tests\test_Seq_objs.py", line 86, in StringMethodTests
    _examples.append(MutableSeq(seq))
  File "C:\projects\biopython\build\lib.win32-3.8\Bio\Seq.py", line 1719, in __init__
    self._data = array.array("u", data._data)
AttributeError: 'UnknownSeq' object has no attribute '_data'
----------------------------------------------------------------------
```

**Comment by peterjc** [comment]
I think it should be:

```
        elif isinstance(data, MutableSeq):
            self._data = # take copy of array
        else:
            # Make no assumptions about the Seq subclass internal storage
            self._data = array.array("u", str(data))
```

**Comment by mdehoon** [comment]
> For the type annotation, rather than:

> `def __init__(self, data: str) -> None`

> we'd need to say data can be `str`,`Seq` or `MutableSeq`.

Sounds good to me. But it goes a bit against Python's duck-typing principle. Also, for fast parsing we will want to add `bytes` and `bytearray`, and if we support lazy parsing then `data` may only be defined by its behavior.

Have a look at `bytes` to see how the Python developers handled this, as it is the closest thing to a `Seq` object.

File: Bio/Seq.py [hunk 2]
```diff
@@ -1700,16 +1705,26 @@ class MutableSeq:
...
+                raise ValueError(
+                    "data should be a string, array of characters, Seq object, "
+                    "or MutableSeq object"
+                )
+            self._data = data
+        elif isinstance(data, str):  # TODO - What about unicode?
             self._data = array.array("u", data)
-        elif isinstance(data, (Seq, int, float)):
+        elif isinstance(data, MutableSeq):
+            self._data = array.array("u", data._data)
```
Comment (peterjc) [inline Bio/Seq.py:1719]: Is this the most efficient way to take a copy of an array? Does this beat it?

```python
self._data = data._data[:] # Take a copy
```

**Review by peterjc — APPROVED** [review]
The changes here fit with the idea of making ``Seq`` etc behave more like the Python string, and remove some non-essential methods.

**Comment by peterjc** [comment]
Sorry, forgot about the else type exception in that snippet:

```python
       elif isinstance(data, MutableSeq):
           self._data = # take copy of array
       elif isinstance(data, Seq):
           # Make no assumptions about the Seq subclass internal storage
           self._data = array.array("u", str(data))
       else:
```

**Comment by mdehoon** [comment]
Thanks!

**Pull request merged**